bn_conversion.c

```diff
@@ -0,0 +1,95 @@
+        while (b != 0u) {                                               \
+            a->dp[i++] = ((mp_digit)b & MP_MASK);                       \
+            if (w <= MP_DIGIT_BIT) { break; }                           \
+            b >>= ((w <= MP_DIGIT_BIT) ? 0 : MP_DIGIT_BIT);             \
+        }                                                               \
+        a->used = i;                                                    \
+        a->sign = MP_ZPOS;                                              \
+        MP_ZERO_DIGITS(a->dp + a->used, a->alloc - a->used);            \
+    }
+#define MP_SET_SIGNED(name, uname, w)                                   \
+    void name(mp_int * a, int##w##_t b)                                 \
+    {                                                                   \
+        uname(a, b < 0 ? -(uint##w##_t)b : (uint##w##_t)b);             \
+        if (b < 0) { a->sign = MP_NEG; }                                \
+    }
+#define MP_INIT_INT(name , set, type)                     \
+    mp_err name(mp_int * a, type b)                      \
+    {                                                    \
+        mp_err err;                                      \
+        if ((err = mp_init(a)) != MP_OKAY) {             \
+            return err;                                  \
+        }                                                \
+        set(a, b);                                       \
+        return MP_OKAY;                                  \
+    }
+#define MP_GET_MAG(name, w)                                             \
+    uint##w##_t name(const mp_int* a)                                   \
+    {                                                                   \
+        unsigned i = MP_MIN((unsigned)a->used, (unsigned)((w + MP_DIGIT_BIT - 1) / MP_DIGIT_BIT)); \
+        uint##w##_t res = 0;                                            \
+        while (i --> 0) {                                               \
+            res <<= ((w <= MP_DIGIT_BIT) ? 0 : MP_DIGIT_BIT);           \
+            res |= (uint##w##_t)a->dp[i];                               \
+            if (w <= MP_DIGIT_BIT) { break; }                           \
+        }                                                               \
+        return res;                                                     \
+    }
+#define MP_GET_SIGNED(name, mag, w)                                     \
+    int##w##_t name(const mp_int* a)                                    \
+    {                                                                   \
+        uint64_t res = mag(a);                                          \
+        return a->sign == MP_NEG ? (int##w##_t)-res : (int##w##_t)res;  \
+    }
+
+#ifdef BN_MP_SET_UINT_C
+MP_SET_UNSIGNED(mp_set_uint, 32)
+#endif
+
+#ifdef BN_MP_SET_UINT64_C
+MP_SET_UNSIGNED(mp_set_uint64, 64)
+#endif
+
+#ifdef BN_MP_SET_SINT_C
+MP_SET_SIGNED(mp_set_sint, mp_set_uint, 32)
```

Comment:
why `sint` and not `int`? IMO it looks and feels wrong

Reply:
Because int is in use already :((( This is the whole point of this PR. 

However I actually started to like sint, since it makes it very obvious that the sign is taken into account!

Reply:
Maybe I should add - before that I used the convention `mp_set_i` and `mp_set_u` , which I think are much uglier.

Reply:
> `mp_set_i` and `mp_set_u` , which I think are much uglier.

that's very true

> Because int is in use already

oops, right, but wait... did we already talk about `mp_set_int32` as possibility? that would also align better with the 64 variant...

> However I actually started to like sint...

first symptoms of the stockholm syndrom? :P

Reply:
> oops, right, but wait... did we already talk about mp_set_int32 as possibility? that would also align better with the 64 variant...

Yes we did. I named things like this to make things feel a bit less heavy during updating. `mp_set_int --> mp_set_uint` instead of `mp_set_int --> mp_set_uint32`. int32 is somehow the "default" and the longer int64 gets the suffix

I could change it and remove the bias towards 32 bits this way if all of you prefer that.

Reply:
What I should say - I somehow tested the API change when I converted test.c and the code base itself. And the names I am using now felt less heavy but still somehow consistent.

The more precise and even more consistent variant would be to add the 32 suffix for sure.

Reply:
> The more precise and even more consistent variant would be to add the 32 suffix for sure.

I think we should go for it

Reply:
@czurnieden ?

Reply:
Fortunately this change is just one sed away

Reply:
I think @czurnieden does not bother.

Reply:
> I think we should go for it

Then we should go for it!

> I think @czurnieden does not bother.

If it isn't way too, uhm, odd I do not care, that is correct.
But you are also right that I would prefer the precise variants, too.